### ISSUE TITLE
MyStudies iOS app does not display published study

### ISSUE THREAD

**Issue by brainsnorkel** [GoogleCloudPlatform/fda-mystudies#3150]
**Describe the bug**

I can't get a study to appear in my MyStudies iOS app.

I have the Google FDA MyStudies  public release 2.0.1 deployed and all workloads are green: sending emails and looking great. I have the iOS app built and handling sign up and sign in. 

I have used Study Builder to create a gateway study that I believe corresponds to my iOS app configuration. I have added sites and participants with Participant Manager and invited them to the published and open for enrolment gateway study.

However, while I can log in to the iOS app (running in the simulator and on my iPhone 11 Pro as a target) the study doesn't appear in the app for me to enter my enrolment token into.

When I launch the app with a MITM proxy the traffic looks normal: I can see HTTP 200 responses for connections to my participant and study endpoints. Yet my study does not appear in the app.

**To Reproduce**

I created a new gateway type study with both iOS and Android app types and published it, updated the app id in the iOS app, updated the gcloud secrets, restarted all the services, and run the copy_app_info_to_sql.sh script. I added a site and invited my own app-user participant but I'm still not getting the study listed in the app.

**Expected behavior**

**Screenshots**

<img width="494" alt="image" src="https://user-images.githubusercontent.com/17165/107459760-f68d4d80-6baa-11eb-99f8-f31654464bbc.png">

**Smartphone (please complete the following information):**
 - Device: iPhone 11 Pro & Simulated iPhone 11 & 12 Pro Max
 - OS: iOS 14.4

**Additional context**

The two strange things I see in the DB tables that I'm not sure are bad:

1. There is a "null" site in the mystudies_participant_datastore locations table that I'm pretty sure I didn't create manually, and
2. There are four entries in the mystudies_participant_datastore  app_info table, one which is the first study/app id I created but the third one seems to have a system-provided default app id where I must have run the copy_app_info_to_sql.sh script without access to gcloud secrets.

For the below excerpts - Note I have created several studies and updated the appid in the app and DB (and seen the study appear in participant manager). The android parameters are placeholder values.

mystudies_participant_datastore contents (elided):

```sql
INSERT INTO `app_info` VALUES ('2c9083827760ce4d01777faee3f20000','woeapp_prod.webofengagement.com','Firebase-Cloud-Messaging-server-key','App Desc_webofengagement','webofengagement','App Name_webofengagement','2021-02-08 03:28:24','0',NULL,NULL,NULL,NULL,'org.chip.woe.mystudies','<<base64 cert p12 elided>>','<<cert password elided>>',NULL,'2021-02-09 00:15:23','0',NULL,NULL),('2c90838c777ff7e20177803f48a60000',NULL,NULL,'App Desc_NOAPPID','NOAPPID','App Name_NOAPPID','2021-02-08 06:06:07','0',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,'2021-02-08 06:06:07',NULL,NULL,NULL),('2c90838d77834b5f017788e018b80000',NULL,NULL,'App Desc_WEBOFENGAGE2','WEBOFENGAGE2','App Name_WEBOFENGAGE2','2021-02-09 22:18:44','0',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,'2021-02-09 22:18:44',NULL,NULL,NULL),('2c90848a7788e20e0177890254b50003','andwoeapp.webofengagement.com','MTI2MjAwMzQ3OTMzQHByb2plY3RzLmdjbS5hbmFTeUIzcmNaTmtmbnFLZEZiOW1oekNCaVlwT1JEQTJKV1d0dw==','App Desc_WEBOFENGAGE3','WEBOFENGAGE3','App Name_WEBOFENGAGE3','2021-02-09 22:56:07','0',NULL,NULL,NULL,NULL,'org.chip.woe.mystudies','<<base64 cert p12 elided>>','<<cert password elided>>',NULL,'2021-02-09 22:56:07',NULL,NULL,NULL);
```

In the DB I see a whole bunch of participants with the same email mapped to different sites and studies:

```sql
INSERT INTO `participant_registry_site` VALUES ('2c90848d778905ed0177891cc93f0000','2021-02-09 23:25:01','c9d30d67-0477-4a8c-8490-0fa1e0300bd0',NULL,'crgentle@gmail.com','YpSXwD9v','2021-02-11 23:25:09',1,'2021-02-09 23:25:09','2021-02-09 23:25:09',NULL,NULL,'I','2c90858b7788e1b8017789030b850006','2c90848a7788e20e0177890254b50004',0),('2c908586777fe93701778007f8300000','2021-02-08 05:05:42','c9d30d67-0477-4a8c-8490-0fa1e0300bd0','2021-02-08 05:18:35','crgentle@gmail.com','nokpFGpq','2021-02-10 05:06:27',2,'2021-02-08 05:06:27','2021-02-08 05:06:27',NULL,NULL,'D','2c908585776ec6d501777fe471fb0000','2c9083827760ce4d01777faee5190001',0),('2c908586777fe9370177801431350006','2021-02-08 05:19:03','c9d30d67-0477-4a8c-8490-0fa1e0300bd0',NULL,'crgentle@gmail.com','QeHVak34','2021-02-10 05:19:56',1,'2021-02-08 05:19:56','2021-02-08 05:19:56',NULL,NULL,'I','2c908586777fe9370177801322920005','2c9083827760ce4d01777faee5190001',0),('2c908586777fe9370177804084c2000a','2021-02-08 06:07:28','c9d30d67-0477-4a8c-8490-0fa1e0300bd0',NULL,'crgentle@gmail.com','eMCgWWW4','2021-02-10 06:07:28',0,NULL,'2021-02-08 06:07:28',NULL,NULL,'N','2c908586777fe937017780401ac50009','2c90838c777ff7e20177803f48c60001',0),('2c90858977834c500177840a5fa50000','2021-02-08 23:46:48','c9d30d67-0477-4a8c-8490-0fa1e0300bd0',NULL,'test@webofengagement.com','na4o48fa','2021-02-10 23:46:59',1,'2021-02-08 23:46:59','2021-02-08 23:46:59',NULL,NULL,'I','2c908586777fe9370177801322920005','2c9083827760ce4d01777faee5190001',0),('2c90858b7788e1b8017788e9ae9f0001','2021-02-09 22:29:12','c9d30d67-0477-4a8c-8490-0fa1e0300bd0',NULL,'crgentle@gmail.com','wckc7DKs','2021-02-11 22:33:21',2,'2021-02-09 22:33:21','2021-02-09 22:33:21',NULL,NULL,'I','2c90858b7788e1b8017788e961fb0000','2c90838d77834b5f017788e019190001',0);
```

It looks like it normalises to the only user email I use for testing: 

```sql
INSERT INTO `user_details` VALUES ('2c9083827760ce4d01777fd828aa0002',NULL,'crgentle@gmail.com',NULL,NULL,NULL,0x01,NULL,'00:00',0x01,NULL,1,0x00,0x00,'f4e3aed2w085c74605ab7efd3490481e1327','2021-02-08 04:13:29','2c9083827760ce4d01777faee3f20000'),('2c90848a7788e20e017788eb79790000',NULL,'crgentle@gmail.com',NULL,NULL,NULL,0x01,NULL,'00:00',0x01,NULL,1,0x01,0x00,'af77e94ek1949z498eb95fdn8c0e5dd2ebd4','2021-02-09 22:31:10','2c90838d77834b5f017788e018b80000');
```

locations

```sql
INSERT INTO `locations` VALUES ('1',NULL,NULL,'location1',NULL,'Y',NULL,NULL,'Site',1),('2c908586777fe93701778012b5ea0004','2021-02-08 05:17:26','c9d30d67-0477-4a8c-8490-0fa1e0300bd0','BCH','The BCH Site','N','2021-02-08 05:17:26',NULL,'Boston Children\'s Hospital',1);
````

**Comment by brainsnorkel** [comment]
I wondered if there were some suggested diagnostics or logs to look for to try to narrow down where the problem is.

**Comment by naveenr-btc** [comment]
Can you confirm if the study is configured for both iOS and Android or just iOS in Study Builder. we feel like it's similar to issue #2850.

**Comment by brainsnorkel** [comment]
Hi Naveen,

It does sound similar and I have pursued the same troubleshooting/remediation as was mentioned in the comments to that issue without luck.

My original study was configured just for iOS, but I have tried a few other approaches since then.

1. I changed that study to add Android and the study still didn't appear in the iOS app.
2. I created a new gateway study with a new app id and configured secrets and the iOS app with the new app id and also set it to have both iOS and Android support in the study (also running the copy_app_info_to_sql.sh script) - still no studies in my iOS app.
 
Regards,
Chris

**Comment by SR-btc** [comment]
Hi, request you to look out for the new master branch release expected to be out tomorrow. This issue could not be reproduced in the latest code and if you could take the updates once it is out tomorrow, that should most likely resolve the issue for you.

**Comment by brainsnorkel** [comment]
Thank you   I’ll take a look then  🙏

**Comment by brainsnorkel** [comment]
Hi Folks,

It doesn't look like master has been merged into yet. Is there branch I should be tracking ahead of the merge?

Regards,
Chris

**Comment by brainsnorkel** [comment]
Oh nevermind. I didn't see the 2.0.2 pull request.

Sorry for the spam!

**Comment by brainsnorkel** [comment]
I have upgraded to 2.0.2 (after some state problems with my local tfswitch using terraform 0.12.30 and cloud build only supporting 0.12.29)

I am still not seeing the studies in the app. I have created a new gateway study, compiled the iOS app with APP_TYPE=gateway and even after signing up and inviting my user I don't see the study in the app.

This is the email I get for the invitation to participate. I wonder if the "App Name_WEBOFENGAGE4" is an indication that something is wrong with my config, as this looks like an auto-generated placeholder and not something I'd want to send to participants.

```
Hi,

You are invited to participate in the Web of Engagement study in the App Name_WEBOFENGAGE4 mobile app.

Please find below the steps to use the enrollment token to enroll into the study using the mobile app:

Download the app > Sign up > Verify your account > Access the study within the app > Enter your unique enrollment token (given below) for the study > Proceed to e-consent and completion of enrollment.

Enrolment Token:   mBIYOyo9

Please note that the token can be used only once and is valid for a period of 48 hours only.

For any questions or assistance, please write to support@webofengagement.com.

Thanks,
The App Name_WEBOFENGAGE4 MyStudies support team
--------------------------------------------------------------------------------
```

Are there troubleshooting steps you'd recommend for narrowing down what my problem should be? Is there an easy way for me to operate the API that should be providing the study list to the app so I can see if there are problems?

**Comment by nishit-v** [comment]
Issue is not reproducible in V2.0.2 for iOS mobile. Assigning to our Developer team for any inputs

**Comment by SR-btc** [comment]
The app name can be updated via this script -https://github.com/GoogleCloudPlatform/fda-mystudies/blob/master/participant-datastore/sqlscript/mystudies_app_info_update_db_script.sql
Future releases will help you manage such app properties via a UI. 

Regarding not seeing studies in the app, could you check if the app has been compiled with the correct App ID value as set up in the Study Builder for the studies you want to see in the app?

**Comment by brainsnorkel** [comment]
Hi SR-btc, 

I have run that script and the APP ID seems to be correct in the iOS app and study builder. I have recreated the gateway study a few times and updated the configs/rebuilt the app - to refer to the APP ID, but there must be something I'm missing. I'll upload a copy of app_info tomorrow morning my time (when I'll recall what I did to export the table last time)

Thanks for your help.

**Comment by brainsnorkel** [comment]
It seems like the app is letting my test user sign up with the right App ID from the information in the in-app registration verification email. 

The APP ID in my APP and in my configuration is "WEBOFENGAGE4" Is the test of the app correct here? 

<img width="859" alt="image" src="https://user-images.githubusercontent.com/17165/109210539-12c2f880-7801-11eb-913d-1b6cb550d794.png">


This is the study in Study Builder:

<img width="848" alt="image" src="https://user-images.githubusercontent.com/17165/109210499-050d7300-7801-11eb-94d7-c56ed8558f69.png">


I don't know why the email has "App Name_WEBOFENGAGE4" - is that a bug, or a sign that there is an uninitialised configuration?

**Comment by brainsnorkel** [comment]
> I have run that script and the APP ID seems to be correct in the iOS app and study builder.

Oops - I commented too soon and misread it as the copy_app_info_to_sql.sh. I had _not_ run that script individually. I'll work on that now.

**Comment by brainsnorkel** [comment]
I have been modifying and debugging the iOS app to ensure that the APP ID is ok and that everything looks ok. 

The APP ID is ok, but It seems that the response I get in the finishedRequest part of the StudyListViewController looks good except it contains no studies.

```javascript
response
(AnyObject?) $R0 = 2 key/value pairs {
  [0] = {
    key = 0x89880304b627e4c6 "message"
    value = 0x8aeb4147b524e726 "SUCCESS"
  }
  [1] = {
    key = 0x88e82385c646f526 "studies"
    value = 0x00007fff8002e970 0 elements
  }
}
```

Is it possible I haven't set up the studies with that APP ID correctly? Is there a step-by-step for setting them up to appear?

**Comment by brainsnorkel** [comment]
I'm wondering if there is a way to test the APIs that provide the study list. It looks like the App is behaving, but my studies are not being delivered to the app. Can anyone help with ideas about how to isolate the problem?

**Comment by aswinijena100** [comment]
> I'm wondering if there is a way to test the APIs that provide the study list. It looks like the App is behaving, but my studies are not being delivered to the app. Can anyone help with ideas about how to isolate the problem?
Yes u can verify the API response  from the postman by invoking the endpoint <domain url>/study-datastore/studyList . Please use IOS specific Authorization (apptoken and bundled) for invoking the API . As per the screen shot attached.

![image](https://user-images.githubusercontent.com/59355719/109627659-d88e8980-7b67-11eb-80ed-8a1c73ea09cc.png)


If you find any error/exception in study datastore console please share with us for further investigation.

**Comment by brainsnorkel** [comment]
That's better. I am getting 401s from the credentials so it's something bad about the way I have configured auth. 

Can you help me with there to get the ios.apptoken value? I can't find it mentioned in the deploy instructions for iOS or the platform and although I have a short alphanumeric string to put here I can't recall where I got that came from or how it is provided to the study-datastore so I can go back and check it for correctness.

These are my GET request headers. If you can check my work I'd appreciate it:

Headers: 

![Screen Shot 2021-03-03 at 12 57 01 pm](https://user-images.githubusercontent.com/17165/109740946-3f6f7980-7c20-11eb-90e0-7bdb4ca4e0c3.png)
applicationId: WEBOFENGAGE4 // This is the value of the manual-mobile-app-appid csecret which matches the gatewaystudy App ID in study builder

Authorization: Basic <<Base64.encode(org.chip.woe.mystudies:\<the ios.apptoken value\>)>>


I have to try to find where I have gone wrong with my auth-related configuration.

**Comment by brainsnorkel** [comment]
I instrumented the study-datastore's StudyMetaDataUtil.java to dump the secrets required for authand tracked down that the auto-sd-ios-id secret is not org.chip.woe.mystudies as I was expecting. 

The value of auto-sd-ios-id in secret manager is a random value generated by mystudies.hcl. 

```
        {
          secret_id   = "auto-sd-ios-id"
          secret_data = "$${random_string.strings[\"sd_ios_id\"].result}"
        },
```

Have I missed a step in the deployment to manualy override this secret value with the package ID?

**Comment by naveenr-btc** [comment]
No you shouldn't replace `auto-sd-ios-id` with package ID.  
The header should be
Authorization: Basic <<Base64.encode(`auto-sd-ios-id`:`the ios.apptoken value`)>> not the value of package ID. 
can you try to config Authorization with this value and try again in postman and let us know the if you still see 401.

**Comment by brainsnorkel** [comment]
That works with a 200 response. The reason I had the package in the iOS app was because of this comment in the Default.xcconfig in the iOS app source that says to use the bundleid:
 
```
// API_KEY is <ios.bundleid>:<ios.apptoken>
// e.g. com.bundleID:axdks
// These values should match the ones set in study datastore server.
// see: study-datastore/src/main/resources/authorizationResource.properties
API_KEY = foo:bar

```

Have I read this comment wrong?

**Comment by brainsnorkel** [comment]
I'm happy to close this cry for help now that I have figured out the information required for the app config. 

I think the documentation could be clearer in the Default.xcconfig comment and the deployment instruction that the first part of the API_KEY should be copied from the output of `gcloud --project=${SECRET_PROJECT} secrets versions access latest --secret=auto-sd-ios-id`

**Comment by aswinijena100** [comment]
Surely we will look into the documentation after discussing with team. However let know if you are able to see the published study in the IOS app , if yes kindly close this ticket .

**Comment by brainsnorkel** [comment]
Studies display, see suggestions for doc and comment updates above.
